src/service-worker-core.js

```diff
@@ -0,0 +1,202 @@
+/**
+ * Copyright 2016 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {startsWith, endsWith} from "./string";
+import indexedDBP from "../third_party/indexed-db-as-promised/index"
+
+const VERSION = '$internalRuntimeVersion$';
+const RELEASE_DATE = versionToDate(VERSION);
+
+function versionToDate(ampVersion) {
+  return new Date(parseInt(ampVersion, 10));
+}
+
+function ampVersion(url) {
+  const matches = /rtv\/(\d+)/.exec(url);
+  return matches ? matches[1] : VERSION;
+}
+
+function isStale(ampVersion, days = 8) {
+  const delta = RELEASE_DATE - new Date(parseInt(ampVersion, 10));
+  return delta > days * (/* 1 day in ms */ 1000 * 60 * 60 * 24);
+}
+
+function basename(url) {
+  return url.substr(url.lastIndexOf('/') + 1);
+}
+
+function versionedUrl(url, version = VERSION) {
+  return url.replace(ampVersion(url), version);
+}
+
+/**
+ * Is this a CDN JS file?
+ * @param {string} url
+ * @return {boolean}
+ */
+function isCdnJsFile(url) {
+  return startsWith(url, 'https://cdn.ampproject.org/') &&
+    endsWith(url, '.js');
+}
+
+
+let cache;
+let db;
+let cacheCleanup = Promise.resolve();
+
+const cachePromise = caches.open('cdn-js').then((result) => {
+  cache = result;
+});
+
+const dbPromise = cachePromise.then(() => {
+  return indexedDBP.open('cdn-js', VERSION, {
+    upgrade(db, { oldVersion, transaction }) {
```

Comment:
are we ok with the closure output, think last time we asked @lannka to remove destructuring